client.go

```diff
@@ -533,88 +515,70 @@ func (c *TunaSessionClient) DialWithConfig(remoteAddr string, config *nkn.DialCo
 		return nil, err
 	}
 
-	var lock sync.Mutex
 	var wg sync.WaitGroup
-	conns := make(map[string]*Conn, len(pubAddrs.Addrs))
-	dialer := &net.Dialer{}
 	for i := range pubAddrs.Addrs {
 		wg.Add(1)
 		go func(i int) {
 			defer wg.Done()
 
-			if pubAddrs.Addrs[i] == nil || len(pubAddrs.Addrs[i].IP) == 0 || pubAddrs.Addrs[i].Port == 0 {
-				return
-			}
-
-			netConn, err := dialer.DialContext(ctx, "tcp", fmt.Sprintf("%s:%d", pubAddrs.Addrs[i].IP, pubAddrs.Addrs[i].Port))
+			_, err = c.dialAConn(ctx, remoteAddr, sessionID, i, pubAddrs.Addrs[i], config)
 			if err != nil {
-				log.Printf("Dial error: %v", err)
+				log.Printf("Tuna session dial to ip %v port %v err: %v", pubAddrs.Addrs[i].IP, pubAddrs.Addrs[i].Port, err)
 				return
 			}
-
-			conn := newConn(netConn)
-
-			err = writeMessage(conn, []byte(c.addr.String()), time.Duration(config.DialTimeout)*time.Millisecond)
-			if err != nil {
-				log.Printf("Write message error: %v", err)
-				conn.Close()
-				return
-			}
-
-			metadata := &pb.SessionMetadata{
-				Id: sessionID,
-			}
-			metadataRaw, err := proto.Marshal(metadata)
-			if err != nil {
-				log.Printf("Encode session metadata error: %v", err)
-				conn.Close()
-				return
-			}
-
-			buf, err := c.encode(metadataRaw, remoteAddr)
-			if err != nil {
-				log.Printf("Encode message error: %v", err)
-				conn.Close()
-				return
-			}
-
-			err = writeMessage(conn, buf, time.Duration(config.DialTimeout)*time.Millisecond)
-			if err != nil {
-				log.Printf("Write message error: %v", err)
-				conn.Close()
-				return
-			}
-
-			lock.Lock()
-			conns[connID(i)] = conn
-			lock.Unlock()
 		}(i)
 	}
 	wg.Wait()
 
+	sessKey := sessionKey(remoteAddr, sessionID)
+	c.RLock()
+	conns := c.sessionConns[sessKey]
+	c.RUnlock()
+	if len(conns) == 0 {
+		return nil, err
+	}
 	connIDs := make([]string, 0, len(conns))
 	for i := 0; i < len(pubAddrs.Addrs); i++ {
 		if _, ok := conns[connID(i)]; ok {
 			connIDs = append(connIDs, connID(i))
 		}
 	}
 
-	sessKey := sessionKey(remoteAddr, sessionID)
 	sess, err := c.newSession(remoteAddr, sessionID, connIDs, config.SessionConfig)
 	if err != nil {
 		return nil, err
 	}
 
 	c.Lock()
 	c.sessions[sessKey] = sess
-	c.sessionConns[sessKey] = conns
 	c.Unlock()
 
 	for i := 0; i < len(pubAddrs.Addrs); i++ {
 		if conn, ok := conns[connID(i)]; ok {
 			go func(conn *Conn, i int) {
 				defer conn.Close()
-				c.handleConn(conn, sessKey, i)
+				for {
+					c.handleConn(conn, sessKey, i)
+
+					if c.config.ReconnectRetries == 0 {
+						return
+					}
+					for j := 0; ; {
```

Comment:
Can be simplified to a single for loop
`j := 0; c.config.ReconnectRetries < 0 || j < c.config.ReconnectRetries; j++`

Reply:
Yes, modified.